src/peer_transport/transport.ts

```diff
@@ -0,0 +1,162 @@
+export const CODE = 276
+
+export interface WebRTCPeerTransportInit {
+  rtcConfiguration?: RTCConfiguration
+}
+
+export interface WebRTCDirectTransportComponents {
+  peerId: PeerId
+  registrar: Registrar
+  upgrader: Upgrader
+  transportManager: TransportManager
+  peerStore: PeerStore
+}
+
+export class WebRTCDirectTransport implements Transport, Startable {
+  private readonly _started = false
+  private readonly handler?: ConnectionHandler
+
+  constructor (
+    private readonly components: WebRTCDirectTransportComponents,
+    private readonly init: WebRTCPeerTransportInit
+  ) {
+    this._onProtocol = this._onProtocol.bind(this)
+  }
+
+  isStarted () {
+    return this._started
+  }
+
+  async start () {
+    await this.components.registrar.handle(PROTOCOL, (data) => {
+      this._onProtocol(data).catch(err => log.error('failed to handle incoming connect from %p', data.connection.remotePeer, err))
+    })
+  }
+
+  async stop () {
+    await this.components.registrar.unhandle(PROTOCOL)
+  }
+
+  createListener (options: CreateListenerOptions): Listener {
+    return new WebRTCPeerListener(this.components)
+  }
+
+  get [Symbol.toStringTag] (): string {
+    return '@libp2p/webrtc-direct'
+  }
+
+  get [symbol] (): true {
+    return true
+  }
+
+  filter (multiaddrs: Multiaddr[]): Multiaddr[] {
+    return multiaddrs.filter((ma) => {
+      const codes = ma.protoCodes()
+      return codes.includes(CODE)
+    })
+  }
+
+  /*
+   * dial connects to a remote via the circuit relay or any other protocol
+   * and proceeds to upgrade to a webrtc connection.
+   * multiaddr of the form: <multiaddr>/webrtc-direct/p2p/<destination-peer>
+   * For a circuit relay, this will be of the form
+   * <relay address>/p2p/<relay-peer>/p2p-circuit/webrtc-direct/p2p/<destination-peer>
+  */
+  async dial (ma: Multiaddr, options: DialOptions): Promise<Connection> {
+    log.trace('dialing address: ', ma)
+    const addrs = ma.toString().split(TRANSPORT)
+    if (addrs.length !== 2) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('invalid multiaddr')
+    }
+    // look for remote peerId
+    const remoteAddr = multiaddr(addrs[0])
+    const destination = multiaddr(addrs[1])
+
+    const destinationIdString = destination.getPeerId()
+    if (destinationIdString == null) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('bad destination')
+    }
+
+    const controller = new AbortController()
+    if (options.signal == null) {
+      options.signal = controller.signal
+    }
+
+    const connection = await this.components.transportManager.dial(remoteAddr)
+
+    const rawStream = await connection.newStream([PROTOCOL], options)
+
+    const pc = await connect({
+      stream: rawStream,
+      rtcConfiguration: this.init.rtcConfiguration,
+      signal: options.signal
+    })
+
+    rawStream.close()
+    void connection.close()
+    // TODO(ckousik): Remove this delay. This is required because
```

Comment:
Consider a case where `A` is connecting to `B` using `webrtc-direct`. `A` and `B` both complete SDP exchange and their PeerConnections transition to the `connected` state. For the ICE connection, `A` is the controlled peer and `B` is the controlling peer. With reference to the description [here](https://datatracker.ietf.org/doc/html/rfc8445#section-7.2.5.3.4), `A` moves to the connected state before `B`. This allows `A` to create a stream on the new connection before `B` has a chance to set up the datachannel callbacks.